### ISSUE TITLE
Reporting color functions as invalid uses of color literals

### ISSUE THREAD

**Issue by chrismbarr** [sasstools/sass-lint#853]
Using `gulp-sass-lint` **v1.2.0**

---

I have this code

``` scss
$color-rgba: rgba(red($input-border-focus), green($input-border-focus), blue($input-border-focus), .6);
box-shadow: $form-shadows, 0 0 8px $color-rgba;
```

Which produces these errors, which seems incorrect since the color names above are [the built-in SCSS functions](http://sass-lang.com/documentation/Sass/Script/Functions.html#red-instance_method), **not** color literals.

```
  27:22  warning  Color functions such as 'red' should not be passed to functions, use variables    no-color-literals
  27:48  warning  Color functions such as 'green' should not be passed to functions, use variables  no-color-literals
  27:76  warning  Color functions such as 'blue' should not be passed to functions, use variables   no-color-literals
```

---

When re-written as a single line like this

``` scss
box-shadow: $form-shadows, 0 0 8px rgba(red($input-border-focus), green($input-border-focus), blue($input-border-focus), .6);;
```

I only get this one error instead

```
  27:39  warning  Color functions such as 'rgba' should only be used in variable declarations  no-color-literals
```

---

Unless I'm missing something this seems like a bug to me, and I don't see a way around this except to either disable this rule (which seems like a bad solution) or to just exclude this one file from linting right now.  

I was able to re-write this code into 4 variables, which is a little bit of an insane work around... but at least it accepts the code now

``` scss
//Due to a bug in `sass-lint` this must be broken up into multiple variables.
//Otherwise it sees the color function as color literals and reports them as violations :/
$color-r: red($input-border-focus);
$color-g: green($input-border-focus);
$color-b: blue($input-border-focus);
$color-rgba: rgba($color-r, $color-g, $color-b, .6);
box-shadow: $form-shadows, 0 0 8px $color-rgba;
```

**Comment by DanPurdy** [comment]
Hmm that is odd, I specifically remember testing for this occurrence, I'll have a look into it.

Thanks for the report.

**Comment by DanPurdy** [comment]
Ok I may have spoken too soon, In all honesty I'm not to sure about whether the rule should be flagging this up or not. Let me have a think about it. 

Also notice that you could just do `rgba($input-border-focus, 0.6)` here as rgba can take just a colour rather than its separate elements.

http://sass-lang.com/documentation/Sass/Script/Functions.html#rgba-instance_method

**Comment by chrismbarr** [comment]
lol, yeah I think you're right, I've got no idea why I wrote the code that way, I just saw this false positive. 

I do think it's still wrong to report this as a failure though.  using the color `red` is not the same as pulling the red value of of an existing color with the function `red(#CCC)`

**Comment by DanPurdy** [comment]
Yeah i agree, I'll look into getting it fixed.

Thanks

**Comment by mattly** [comment]
It does this for key names in a map as well, for example:

``` sass
$colors: {
  black: #1A1A1A,
}
```

flags `color 'black' should be written in its hexadecimal form #000000`

**Comment by DanPurdy** [comment]
@mattly That's one of the options you can enable / disable

https://github.com/sasstools/sass-lint/blob/develop/docs/rules/no-color-literals.md#options

**Comment by wawjr3d** [comment]
i see a related issue with color-literals when I write:

```css
background-color: rgba(shade($tomato, 20), .8);
```

I get the warning/error `A color in variable form must be passed to rgba, literals are restricted  no-color-literals`. There's no error when I remove the shade function call:

```css
background-color: rgba($tomato, .8);
```

**Comment by alexmccabe** [comment]
I'm having issues getting sass-lint to recognise these options. I can binary turn the `no-color-literals` on and off, but if I add `- allow-map-identifiers: 1` as an option, nothing changes. Additionally, if I add `allow-map-identifiers: 1` at the root level of `rules` the entire sass-lint.yml doesn't get loaded. 

Can you point me in the direction of where to set this config? It's kinda irritating that this gets flagged. I'd rather not turn the rule off entirely. 

It also flags if I do `map-get(blue, light)`. Which is SUPER frustrating. 

_Note: I have come over from `scss-lint` as that is being end-of-life'd. I used the converter tool provided, and got my basic setup how I like it. Having this would just be a lovely cherry on top._

**Comment by chrismbarr** [comment]
Just found another related thing today.

With this code: `rgba($color, .5)` I get this linting violation:
```
Color functions such as 'rgba' should not be passed to functions, use variables  no-color-literals
```

I think the problem is that the linter cannot tell the difference between the native CSS `rgba` color literal and [the `rgba` SCSS color function](http://sass-lang.com/documentation/Sass/Script/Functions.html#rgba-instance_method)

**Desired change:**
`rgba(255, 255, 255, .5);` - This **should** still be a violation of the `no-color-literals` rule
`rgba($color, .5)` - This is an SCSS function and not a color literal, it should not trigger a linting violation

**Comment by blueocto** [comment]
Just to +1 what @chrismbarr has posted, I'm trying out sass-lint for the first time and getting setup with my gulpfile options ... however I'm also getting a warning for 

> warning  Color functions such as 'rgba' should only be used in variable declarations    no-color-literals

When my code actually is a variable;
`border-bottom: 1px solid rgba($light-gray, .5); `

Ideally would not like to turn this who feature off, just perhaps the part that flag's this particular warning.

**Comment by chrismbarr** [comment]
@blueocto As a workaround for now I've found that you either need to disable the rule with an explanation like this
```scss
.some-element {
    //sass-lint:disable-block no-color-literals  --  This is needed to get around a bug in sass-lint until it's fixed
    //https://github.com/sasstools/sass-lint/issues/853
    border-bottom: 1px solid rgba($light-gray, .5);
}
```

or switch to use a similar function like [`opacify`](http://sass-lang.com/documentation/Sass/Script/Functions.html#opacify-instance_method) or [`transparentize`](http://sass-lang.com/documentation/Sass/Script/Functions.html#transparentize-instance_method)

```scss
.some-element {
    border-bottom: 1px solid transparentize($light-gray, .5);
}
```

**Comment by DanPurdy** [comment]
@chrismbarr @blueocto both of your recent examples work and are not flagged if you enable the `allow-rgba` option for the rule. See [here](https://github.com/sasstools/sass-lint/blob/develop/docs/rules/no-color-literals.md#allow-rgba-true).

```
.some-element {
    border-bottom: 1px solid rgba($light-gray, .5);
}
```

and 

```
border-bottom: 1px solid rgba($light-gray, .5);
```

both report no lint errors with this enabled.

**Comment by DanPurdy** [comment]
@alexmccabe I think you've got an invalid config file there, it should work with this set

```yml
no-color-literals:
    - 1
    -
      allow-map-identifiers: true
```

notice the return after the dash for the extra options and also the string true as defined in the [docs](https://github.com/sasstools/sass-lint/blob/develop/docs/rules/no-color-literals.md#allow-map-identifiers-false)

Try validating your config with http://www.yamllint.com/ too just in case. There's a PR i started ages ago and need to finish to allow json configs also as I don't find YAML that nice to use either..

**Comment by DanPurdy** [comment]
I'm going to add a fix to make sure you can pass functions and variables to rgba, definitely an oversight why that was disabled in the first place...

**Comment by alexmccabe** [comment]
@DanPurdy Ah ace! Thanks for that, the Yaml syntax is just rather odd, so a JSON config would be much preferred. JSON is king! 

As for the other example (`map-get(blue, light)` flagging), the only thing I can think of is to check whether the colour is inside of a select few SCSS functions. However, this is incredibly unclean and probably wouldn't want to be the maintainer of such a list, as it will never please anyone.

**Comment by DanPurdy** [comment]
Released in 1.12.0

**Comment by chrismbarr** [comment]
Glad this is fixed now, I haven't gotten a chance to try it out yet though.  I just wanted to drop in my last $0.02 here.  Having an option like `allow-rgba` seems odd to me.  I can't really see a lot of situations where you'd want this to be off (which is the default right now).  IMO, this option should just be enabled be default.

**Comment by DanPurdy** [comment]
@chrismbarr yeah i totally agree, I think we (read me) went a bit too far with the configurability of this. If you want to open another issue with maybe your thoughts on the options that would be useful for this rule and maybe it's default behaviour then we could look to get it into v2 or v3 of which v2 is going to land soon as we've got so much we want to change.

**Comment by m10** [comment]
`color: rgba($color-black, .5);` still gives me a warning with 1.12.1

**Comment by Jaason** [comment]
The same here: `box-shadow: 0 6px 12px rgba($black, .175)`

**Comment by UisgeBeatha436** [comment]
Hi, 
I'm facing similar warnings: _Color functions such as 'rgba' should not be passed to functions, use variables._

```
$bgdefault: #1a1a3f;
@mixin bg-gradient { // ()
  transition: all .5s;
  // background: rgb(187, 194, 194);
  background: linear-gradient(90deg, rgba(187, 194, 194, 1) 0%, rgba(1, 72, 136, 1) 35%, rgba(26, 26, 63, 1) 100%);
}
```
any suggestion on I an solve this are much appreciated.
